src/RefinedAst.hs

```diff
@@ -22,14 +22,25 @@ import Data.Aeson.Types
 import Data.Vector (fromList)
 
 -- AST post typechecking
-data Claim = B Behaviour | I Invariant | S Storages deriving (Show, Eq)
+data Claim = C Constructor | B Behaviour | I Invariant | S Store deriving (Show, Eq)
+
+type Store = Map Id (Map Id SlotType)
 
-newtype Storages = Storages { unStorages :: Map Id (Map Id SlotType) } deriving (Show, Eq)
 data Invariant = Invariant Id (Exp Bool) deriving (Show, Eq)
+
+data Constructor = Constructor
+  { _cname :: Id,
+    _cmode :: Mode,
+    _cinterface :: Interface,
+    _cpreconditions :: [Exp Bool],
+    _cpostconditions :: [Exp Bool],
+    _initialStorage :: [StorageUpdate],
+    _cstateUpdates :: [Either StorageLocation StorageUpdate]
```

Comment:
Initialstorage is the initial values (and definition) of the contracts storage being created and `cstateupdatrs` is other contraxt. Im not sure if the distinctions is a good idea. Wdyt?

Reply:
What is the benefit of seperating the two?

Reply:
a stronger guarantee that the `initialState` of a contract must consist of a `StorageUpdate` instead of being `Either StorageLocation StorageUpdate`
On the other hand there is nothing in the type that guarantees that the initialState stems from the contract in question anyway, so maybe it's a little bit of a lost cause to begin with